src/server.rs

```diff
@@ -0,0 +1,409 @@
+    // telling them what's inside.
+    // https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs
+    _private: [u8; 0],
+}
+
+#[allow(non_camel_case_types)]
+pub struct rustls_server_session {
+    _private: [u8; 0],
+}
+
+/// Create a rustls_server_config_builder. Caller owns the memory and must
+/// eventually call rustls_server_config_builder_build, then free the
+/// resulting rustls_server_config. This starts out with no trusted roots.
+/// Caller must add roots with rustls_server_config_builder_load_native_roots
+/// or rustls_server_config_builder_load_roots_from_file.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.ServerConfig.html#method.new
+#[no_mangle]
+pub extern "C" fn rustls_server_config_builder_new() -> *mut rustls_server_config_builder {
+    ffi_panic_boundary_ptr! {
+        let config = rustls::ServerConfig::new(Arc::new(NoClientAuth));
+        let b = Box::new(config);
+        Box::into_raw(b) as *mut _
+    }
+}
+
+/// Sets a single certificate chain and matching private key.
+/// This certificate and key is used for all subsequent connections,
+/// irrespective of things like SNI hostname.
+/// cert_chain must point to a byte array of length cert_chain_len containing
+/// a series of PEM-encoded certificates, with the end-entity certificate
+/// first.
+/// private_key must point to a byte array of length private_key_len containing
+/// a private key in PEM-encoded PKCS#8 or PKCS#1 format.
+#[no_mangle]
+pub extern "C" fn rustls_server_config_builder_set_single_cert_pem(
+    builder: *mut rustls_server_config_builder,
+    cert_chain: *const u8,
+    cert_chain_len: size_t,
+    private_key: *const u8,
+    private_key_len: size_t,
+) -> rustls_result {
+    ffi_panic_boundary! {
+        let config: &mut ServerConfig = try_ref_from_ptr!(builder, &mut ServerConfig);
+        let mut cert_chain: &[u8] = unsafe {
+            if cert_chain.is_null() {
+                return NullParameter;
+            }
+            slice::from_raw_parts(cert_chain, cert_chain_len as usize)
+        };
+        let private_key: &[u8] = unsafe {
+            if private_key.is_null() {
+                return NullParameter;
+            }
+            slice::from_raw_parts(private_key, private_key_len as usize)
+        };
+        let mut private_keys: Vec<PrivateKey> = match pkcs8_private_keys(&mut Cursor::new(private_key)) {
+            Ok(v) => v,
+            _ => match rsa_private_keys(&mut Cursor::new(private_key)) {
+                Ok(v) => v,
+                Err(()) => return rustls_result::PrivateKeyParseError,
+            }
+        };
+        let private_key: PrivateKey = match private_keys.pop() {
+            Some(p) => p,
+            None => return rustls_result::PrivateKeyParseError,
+        };
+        let parsed_chain: Vec<Certificate> = match certs(&mut cert_chain) {
+            Ok(v) => v,
+            Err(()) => return rustls_result::CertificateParseError,
```

Comment:
Oof, [`Result<Vec<Certificate>, ()>`](https://docs.rs/rustls/0.19.0/rustls/internal/pemfile/fn.certs.html) is a strange return type.

Reply:
[`rusttls_pemfile::certs`](https://docs.rs/rustls-pemfile/0.2.0/rustls_pemfile/fn.certs.html) is a little nicer.